### PR TITLE
➕Add warnings ⚠️ for inproper task setting in config files.

### LINKED ISSUE #273: Test anomalib with unsupervised dataset
I'm trying to adapt the current pipeline for some ad hoc dataset (ie EuroSAT) in order to test anomalib and make it more robust. For using the `FolderDataset` for unsupervised segmentation task.

So I'm using the EuroSAT RGB Dataset (https://madm.dfki.de/files/sentinel/EuroSAT.zip), Residential as normal and Industrial as abnormal, and of course no mask/groundtruth.

As everything is fine now on development branch with torchmetrics, I will use 24ccb585319dd10ef81778364350eb972bf1cf41.

# Updating config with EuroSAT

Here is my dataset section I'm adding using to `cflow`,  `ganomaly`, `padim` and `patchcore` (others settings are not modified)

> ```yaml
> dataset:
>   name: eurosat
>   format: folder
>   path: ./datasets/EuroSAT
>   normal_dir: Residential
>   abnormal_dir: Industrial
>   task: segmentation
>   normal_test_dir: null
>   mask: null
>   extensions: .jpg
>   split_ratio: 0.2
>   seed: 0
>   image_size: 64
>   train_batch_size: 16
>   test_batch_size: 16
>   num_workers: 4
>   transform_config:
>     train: null
>     val: null
>   create_validation_set: false
>   tiling:
>     apply: false
>     tile_size: null
>     stride: null
>     remove_border_count: 0
>     use_random_tiling: False
>     random_tile_count: 16
>   inference_batch_size: 16
>   fiber_batch_size: 32 #for padim
>   ```
## Cflow: metric `pixel_AUROC` not available

> ```
> Epoch 0: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 344/344 [04:10<00:00,  1.38it/s]Traceback (most recent call last):                                                                                                                                              
>   File "tools/train.py", line 83, in <module>
>     train()
>   File "tools/train.py", line 73, in train
>     trainer.fit(model=model, datamodule=datamodule)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 768, in fit
>     self._call_and_handle_interrupt(
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 721, in _call_and_handle_interrupt
>     return trainer_fn(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 809, in _fit_impl
>     results = self._run(model, ckpt_path=self.ckpt_path)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1234, in _run
>     results = self._run_stage()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1321, in _run_stage
>     return self._run_train()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1351, in _run_train
>     self.fit_loop.run()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 205, in run
>     self.on_advance_end()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 297, in on_advance_end
>     self.trainer._call_callback_hooks("on_train_epoch_end")
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1634, in _call_callback_hooks
>     fn(self, self.lightning_module, *args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/callbacks/early_stopping.py", line 179, in on_train_epoch_end
>     self._run_early_stopping_check(trainer)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/callbacks/early_stopping.py", line 190, in _run_early_stopping_check
>     if trainer.fast_dev_run or not self._validate_condition_metric(  # disable early_stopping with fast_dev_run
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/callbacks/early_stopping.py", line 145, in _validate_condition_metric
>     raise RuntimeError(error_msg)
> RuntimeError: Early stopping conditioned on metric `pixel_AUROC` which is not available. Pass in or modify your `EarlyStopping` callback to use any of the following: `image_F1Score`, `image_AUROC`
> ```

Change early_stopping > `metric: pixel_AUROC` to `metric: image_AUROC`. :heavy_check_mark:

## Padim: `KeyError: 'mask'`

> ```
> Epoch 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 375/375 [00:14<00:00, 25.17it/s, loss=nan]
> Training took 14.995270490646362 seconds                                                                                                                                        
> LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
> Testing DataLoader 0:   0%|                                                                                                                             | 0/225 [00:00<?, ?it/s]Traceback (most recent call last):
>   File "tools/train.py", line 83, in <module>
>     train()
>   File "tools/train.py", line 79, in train
>     trainer.test(model=model, datamodule=datamodule)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 936, in test
>     return self._call_and_handle_interrupt(self._test_impl, model, dataloaders, ckpt_path, verbose, datamodule)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 721, in _call_and_handle_interrupt
>     return trainer_fn(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 983, in _test_impl
>     results = self._run(model, ckpt_path=self.ckpt_path)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1234, in _run
>     results = self._run_stage()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1318, in _run_stage
>     return self._run_evaluate()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1363, in _run_evaluate
>     eval_loop_results = self._evaluation_loop.run()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 204, in run
>     self.advance(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/dataloader/evaluation_loop.py", line 153, in advance
>     dl_outputs = self.epoch_loop.run(self._data_fetcher, dl_max_batches, kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 204, in run
>     self.advance(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 133, in advance
>     self._on_evaluation_batch_end(output, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 263, in _on_evaluation_batch_end
>     self.trainer._call_callback_hooks(hook_name, output, *kwargs.values())
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1634, in _call_callback_hooks
>     fn(self, self.lightning_module, *args, **kwargs)
>   File "/home/gsd/Desktop/julien-blanchon/trash/anomalib/anomalib/utils/callbacks/visualizer_callback.py", line 141, in on_test_batch_end
>     true_mask = outputs["mask"][i].cpu().numpy() * 255
> KeyError: 'mask'
> Testing DataLoader 0:   0%|          | 0/225 [00:00<?, ?it/s]       
> ```

I get an `KeyError: 'mask'` after the first epoch, as ground truth is not provide the mask visualization step of on_batch_end must be disable. https://github.com/openvinotoolkit/anomalib/blob/24ccb585319dd10ef81778364350eb972bf1cf41/anomalib/utils/callbacks/visualizer_callback.py#L140 
After a small fix (ie skip ground truth visualization) everything work fine.  :heavy_check_mark:

```python
#TODO: Make true_mask visualisation optional as ground thruth is not always provide 
if self.task == "segmentation" and "mask" in outputs:
   true_mask = outputs["mask"][i].cpu().numpy() * 255
   visualizer.add_image(image=true_mask, color_map="gray", title="Ground Truth")
```

## Patchcore

> ```
> Traceback (most recent call last):███████████████████████████████████████████████████████████████████████████████████████| 194/194 [01:23<00:00,  2.32it/s]
>   File "tools/train.py", line 83, in <module>
>     train()
>   File "tools/train.py", line 73, in train
>     trainer.fit(model=model, datamodule=datamodule)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 768, in fit
>     self._call_and_handle_interrupt(
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 721, in _call_and_handle_interrupt
>     return trainer_fn(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 809, in _fit_impl
>     results = self._run(model, ckpt_path=self.ckpt_path)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1234, in _run
>     results = self._run_stage()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1321, in _run_stage
>     return self._run_train()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1351, in _run_train
>     self.fit_loop.run()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 204, in run
>     self.advance(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 269, in advance
>     self._outputs = self.epoch_loop.run(self._data_fetcher)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 205, in run
>     self.on_advance_end()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 255, in on_advance_end
>     self._run_validation()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 309, in _run_validation
>     self.val_loop.run()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 211, in run
>     output = self.on_run_end()
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/dataloader/evaluation_loop.py", line 187, in on_run_end
>     self._evaluation_epoch_end(self._outputs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/loops/dataloader/evaluation_loop.py", line 309, in _evaluation_epoch_end
>     self.trainer._call_lightning_module_hook("validation_epoch_end", output_or_outputs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1593, in _call_lightning_module_hook
>     output = fn(*args, **kwargs)
>   File "/home/gsd/Desktop/julien-blanchon/trash/anomalib/anomalib/models/components/base/anomaly_module.py", line 133, in validation_epoch_end
>     self._collect_outputs(self.image_metrics, self.pixel_metrics, outputs)
>   File "/home/gsd/Desktop/julien-blanchon/trash/anomalib/anomalib/models/components/base/anomaly_module.py", line 159, in _collect_outputs
>     image_metric.update(output["pred_scores"], output["label"].int())
>   File "/home/gsd/Desktop/julien-blanchon/trash/anomalib/anomalib/utils/metrics/collection.py", line 37, in update
>     super().update(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/collections.py", line 153, in update
>     m.update(*args, **m_kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/metric.py", line 312, in wrapped_func
>     update(*args, **kwargs)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/classification/stat_scores.py", line 185, in update
>     tp, fp, tn, fn = _stat_scores_update(
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/functional/classification/stat_scores.py", line 154, in _stat_scores_update
>     preds, target, _ = _input_format_classification(
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/utilities/checks.py", line 407, in _input_format_classification
>     case = _check_classification_inputs(
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/utilities/checks.py", line 268, in _check_classification_inputs
>     case, implied_classes = _check_shape_and_type_consistency(preds, target)
>   File "/home/gsd/anaconda3/envs/anomalib_env/lib/python3.8/site-packages/torchmetrics/utilities/checks.py", line 114, in _check_shape_and_type_consistency
>     raise ValueError(
> ValueError: Either `preds` and `target` both should have the (same) shape (N, ...), or `target` should be (N, ...) and `preds` should be (N, C, ...).
> Epoch 0: 100%|██████████| 344/344 [01:44<00:00,  3.29it/s, loss=nan]   
> ```

Shape error :(, any idea ? :x:

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- This PR adds checks to see if the task is properly assigned in folder dataset.

- Fixes #273 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/data/folder.py [hunk 1]
```diff
@@ -221,11 +222,18 @@ def __init__(
...
+                "Classification is to be chosen if mask directory is not provided."
+            )
+
         if task == "classification" and mask_dir:
-            raise ValueError(
+            warnings.warn(
                 "Classification task is requested, but mask directory is provided. "
                 "Segmentation task is to be chosen if mask directory is provided."
             )
+
```
Comment (ashwinvaidya17) [inline anomalib/data/folder.py:236]: Likewise, `self.task = "segmentation"`? or is it being assigned elsewhere?

**Review by ashwinvaidya17 — APPROVED** [review]
I think this will also address the comment on the issue about classification task producing only binary predictions rather than anomaly masks.

**Review by djdameln — APPROVED** [review]
I guess the task type parameter really is redundant for the folder dataset. Especially now that performance metrics are configurable. The user can now choose between classification and segmentation evaluation in 3 ways: `task_type` parameter, `mask_path` parameter and `metrics.pixel` parameter. In understand why this confuses the users, and I feel we need a better design for this. But let's merge this PR as a quick fix and create a better design separately.

**Comment by samet-akcay** [comment]
I agree it's not so intuitive. We need to make it simpler.

**Pull request merged**